test-framework/core/src/main/java/org/keycloak/testframework/realm/UserSupplier.java

```diff
@@ -41,12 +46,33 @@ public ManagedUser getValue(InstanceContext<ManagedUser, InjectUser> instanceCon
             if (Status.CONFLICT.equals(Status.fromStatusCode(response.getStatus()))) {
                 throw new IllegalStateException("User already exist with username: " + userRepresentation.getUsername());
             }
-            String uuid = ApiUtil.handleCreatedResponse(response);
+            String userUuid = ApiUtil.handleCreatedResponse(response);
+
+            instanceContext.addNote(USER_UUID_KEY, userUuid);
 
-            instanceContext.addNote(USER_UUID_KEY, uuid);
+            UserResource userResource = realm.admin().users().get(userUuid);
+            userRepresentation.setId(userUuid);
 
-            UserResource userResource = realm.admin().users().get(uuid);
-            userRepresentation.setId(uuid);
+            Map<String, List<String>> clientRoles = userRepresentation.getClientRoles();
+            if (clientRoles == null || clientRoles.isEmpty()) {
+                return new ManagedUser(userRepresentation, userResource);
+            }
+            // replace map keys (instead of clientId we need the client's uuid)
+            Map<String, List<String>> clientRolesUuid = clientRoles.entrySet().stream().collect(Collectors.toMap(
+                    entry -> realm.admin().clients().findByClientId(entry.getKey()).stream()
+                            .findFirst()
+                            .orElseThrow()
+                            .getId(),
+                    Map.Entry::getValue
+            ));
+            // replace the string described roles with actual role representations
+            Map<String, List<RoleRepresentation>> clientRolesRepresentations = clientRolesUuid.entrySet().stream().collect(Collectors.toMap(
+                    Map.Entry::getKey,
+                    entry -> realm.admin().users().get(userUuid).roles().clientLevel(entry.getKey()).listAvailable().stream()
```

Comment:
Just quickly spotted - the user should be obtained before obtaining the `clientRolesRepresentations`.

Reply:
Oh, right, so it doesn't make that many requests, right? That reminds me that `UserResource` is available from a few lines above. I even use it below. So the line can be shortened to:
```java
entry -> userResource.roles().clientLevel(entry.getKey()).listAvailable().stream()
```